[Metricbeat] CoreDNS module: Add Kibana Dashboard

ioandr opened the pull request:
This PR adds an overview dashboard for the CoreDNS Metricbeat module as proposed on #10432.

This is the first iteration on this dashboard, so please share any thoughts on the proposed visualizations.

![coredns-dashboard-p1](https://user-images.githubusercontent.com/14974598/55441696-65172280-55b4-11e9-9c55-59a9ec15a019.png)

![coredns-dashboard-p2](https://user-images.githubusercontent.com/14974598/55441831-e40c5b00-55b4-11e9-8c45-6aeee534e078.png)

![coredns-dashboard-p3](https://user-images.githubusercontent.com/14974598/55441746-9394fd80-55b4-11e9-83ee-8ac1f608a4f4.png)

Signed-off-by: Ioannis Androulidakis <ioannis@arrikto.com>

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

ChrsMark commented:
Hello @ioandr @ruflin ,

hopping into this to share a though.

The Prometheus metrics we collect and try to visualize here that are of cumulative-histograms type should be ideally visualized with a heatmap. Prometheus histograms are nicely visualized like this in Grafana:
![GrafanaHeatmap](https://user-images.githubusercontent.com/11754898/55711914-643c2180-59f6-11e9-9e34-97f6cbc090b0.jpg)

A nice blog describing this: [Histograms to the rescue]( https://blog.cloudflare.com/introducing-ebpf_exporter/ )

So, is there a similar way to create such visualizations in Kibana and make the most out of there CoreDNS metrics (size_bytes, response_ns etc)?

odacremolbap commented:
Hi @ioandr @ChrsMark 

So much thanks for your contributions, and for your patience on this PR.

https://github.com/elastic/beats/pull/11738 was merged which brought some renaming + new metrics for the CoreDNS cache, and renaming fields to conform to [this](https://www.elastic.co/guide/en/beats/devguide/current/event-conventions.html#abbreviations)

Can you please rebase, and adapt to the fields renaming?

Thanks!

ioandr commented:
> Hi @ioandr @ChrsMark
> 
> So much thanks for your contributions, and for your patience on this PR.
> 
> #11738 was merged which brought some renaming + new metrics for the CoreDNS cache, and renaming fields to conform to [this](https://www.elastic.co/guide/en/beats/devguide/current/event-conventions.html#abbreviations)
> 
> Can you please rebase, and adapt to the fields renaming?
> 
> Thanks!

Hey @odacremolbap, thanks for the heads-up on this!

I will take a look at the new CoreDNS metrics and field names and update this PR accordingly.

exekias commented:
ey @ioandr, did you have a chance to have look to this? it would be nice to get this in

ioandr commented:
> ey @ioandr, did you have a chance to have look to this? it would be nice to get this in

Hi @exekias, I finally found the time to make some progress on this, thanks for your reminder.

@odacremolbap I renamed certain metrics as well as names of corresponding visualizations. Also, I added another visualization with cache-related metrics in the bottom of the dashboard. Let me know about your opinion on these changes, so we can have another round of review.

Please note that I am taking some days off, roughly until May 4th, so I won't be available until then. Feel free to commit on this PR so I do not block anything!

odacremolbap commented:
thanks a lot @ioandr 
taking a look

ioandr commented:
> thanks a lot @ioandr
> taking a look

Hi @odacremolbap, any update on this? Please let me know If there is anything I can do to facilitate your review.

odacremolbap commented:
@ioandr 
I'm having a couple of issues 
- usage of non keywords as terms for aggregation
- usage of data that is not being informed (in my case `coredns.stats.dns.request.do.count`) present as aggregation. Although a filter is present it looks like Kibana is evaluating and popping up an error.

It would help to know which version of ES/Kibana did you use so that I configure mine the same.

Thanks!

odacremolbap commented on metricbeat/module/coredns/_meta/kibana/7/dashboard/Metricbeat-coredns-overview.json:
```diff
@@ -0,0 +1,1395 @@
+            }
+          }
+        },
+        "title": "Requests by Protocol [Metricbeat CoreDNS] ECS",
```
This visualization is counting documents.
By what I can read the data comes by `family` `proto` `server` `zone`, and the field contains the total count of requests.

The pie should
- show the number (top hit I guess, since this is incremental) instead of the count of docs
- the number at the metrics endpoint is divided into server/proto/zone/family. We need to account for that, I tested aggregations bucketing on zone, then on protocol, and the 2 concentric circles look much more accurate. Then there are families (I haven't tested ipv6, can we stick for now to family=1 and not support ipv6?). Probably we can split charts by `server` in case someone configures more than one

WDYT?

odacremolbap commented on metricbeat/module/coredns/_meta/kibana/7/dashboard/Metricbeat-coredns-overview.json:
```diff
@@ -0,0 +1,1395 @@
+            }
+          }
+        },
+        "title": "Requests by Type [Metricbeat CoreDNS] ECS",
```
this metric is also provided by server, type and zone. Instead of getting the average of those different counts, we can do some _per zone splitting_ (I think that will also remove the `success` type which comes informed from a set of metrics)

odacremolbap commented on metricbeat/module/coredns/_meta/kibana/7/dashboard/Metricbeat-coredns-overview.json:
```diff
@@ -0,0 +1,1395 @@
+            }
+          }
+        },
+        "title": "Request Duration [Metricbeat CoreDNS] ECS",
```
these metrics seem to be stacked, but their values are accumulative. Buckets contains counts for all requests under their specified duration, so bucket `+Inf` contains all counts.

Probably showing as line or area, without stacking data would be better.

Again, this data seems to be informed by `server` and `zone`. Splitting would be desired.

odacremolbap commented on metricbeat/module/coredns/_meta/kibana/7/dashboard/Metricbeat-coredns-overview.json:
```diff
@@ -0,0 +1,1395 @@
+            }
+          }
+        },
+        "title": "Request Size Bytes [Metricbeat CoreDNS] ECS",
```
pretty much like the request duration case.
This time metric is provided by `proto`, `server`, and `zone`
Averaging different values can be misleading

odacremolbap commented on metricbeat/module/coredns/_meta/kibana/7/dashboard/Metricbeat-coredns-overview.json:
```diff
@@ -0,0 +1,1395 @@
+            }
+          }
+        },
+        "title": "Requests by Zone [Metricbeat CoreDNS] ECS",
```
you can use the `coredns.stats.dns.request.count` instead of the document count to show this metrics.
Also, this metric is informed by `family`, `proto, `server` and `zone`.

odacremolbap commented on metricbeat/module/coredns/_meta/kibana/7/dashboard/Metricbeat-coredns-overview.json:
```diff
@@ -0,0 +1,1395 @@
+            }
+          }
+        },
+        "title": " Cache Hits, Misses [Metricbeat CoreDNS] ECS",
```
the metric being shown is an always increasing count. What about using Max instead of Avg?

odacremolbap commented:
@ioandr 
I think my requested changes might need a decent amount of work and testing. I might contribute to this branch and sync with you at this thread, if you don't mind.

ioandr commented:
> @ioandr
> I think my requested changes might need a decent amount of work and testing. I might contribute to this branch and sync with you at this thread, if you don't mind.

Sure, feel free to contribute on this PR. I will take a look at the requested changes and come back

odacremolbap commented:
I've added per server/zone/family/proto metrics where needed.

Since prometheus endpoints return a cumulative number (total counts), graphs looks boring.
I'm not using bucketed metrics at all, will dive in here some more, probably a data table panel would be nice to add.

odacremolbap commented:
there are some issues with my visualizations, working on it

kaiyan-sheng commented:
jenkins, test this please

kaiyan-sheng commented:
I was able to load the dashboard, but I'm having issue running coreDNS so I didn't get any data(my problem).

odacremolbap commented:
@ioandr your PR got merged.
Thanks a lot for the contribution!
